### ISSUE TITLE
wrong sort inferred

### ISSUE THREAD

**Issue by grosu** [kframework/k-legacy#605]
Took me half an hour to debug `tutorial/1_k/5_types/lesson_3` to find out that the sort inference algorithm inferred the wrong sort of `E` in this rule:

```
  rule <k> lambda X : T . E => (T -> E) ~> tenv(Rho) ...</k>
       <tenv> Rho:Map => Rho[T/X] </tenv>
```

For some reason, it inferred that the sort of `E` is `Type` instead of `Exp`.  I had to look into the generated Maude file to find this error, which sucks; we need a way to quickly see what sorts were inferred for the variables in a given rule, as well as why a certain rule does not apply.  But that is another issue, which I will report next.

For your reference, here is the signature:

```
  syntax Type ::= "int" | "bool"
                | Type "->" Type
                | "(" Type ")"                 [bracket]

  syntax Exp ::= Id
               | "lambda" Id ":" Type "." Exp
               | Exp Exp                       [strict, left]
               | "(" Exp ")"                   [bracket]

  syntax Exp ::= Type

  syntax KResult ::= Type

  syntax Exp ::= Exp "->" Exp                  [strict]
  rule <k> lambda X : T . E => (T -> E) ~> tenv(Rho) ...</k>
       <tenv> Rho:Map => Rho[T/X] </tenv>
```

p.s. The rule above has other problems, too, but let's not worry about those here.

**Comment by radumereuta** [comment]
This is related to #109. So, @grosu please respond after you read the older issue.

**Comment by grosu** [comment]
I just read #109 and answered it.  According to the first principle there, you would infer the sort Exp for E in this example.

**Comment by radumereuta** [comment]
Should be fixed with #698.
